File: app/src/main/java/io/fourth_finger/scrabble/dictionary/LetterNode.kt
package io.fourth_finger.scrabble.dictionary

class LetterNode private constructor(
    val char: Char,
    val score: Int
){

    override fun toString(): String {
        return "$char: $score"
    }

    companion object {
        private val SPACE = LetterNode(' ', 0)

        private val E = LetterNode('E', 1)
        private val A = LetterNode('A', 1)
        private val I = LetterNode('I', 1)
        private val O = LetterNode('O', 1)
        private val N = LetterNode('N', 1)
        private val R = LetterNode('R', 1)
        private val T = LetterNode('T', 1)
        private val L = LetterNode('L', 1)
        private val S = LetterNode('S', 1)
        private val U = LetterNode('U', 1)

        private val D = LetterNode('D', 2)
        private val G = LetterNode('G', 2)

        private val B = LetterNode('B', 3)
        private val C = LetterNode('C', 3)
        private val M = LetterNode('M', 3)
        private val P = LetterNode('P', 3)

        private val F = LetterNode('F', 4)
        private val H = LetterNode('H', 4)
        private val V = LetterNode('V', 4)
        private val W = LetterNode('W', 4)
        private val Y = LetterNode('Y', 4)

        private val K = LetterNode('K', 5)

        private val J = LetterNode('J', 8)
        private val X = LetterNode('X', 8)

        private val Q = LetterNode('Q', 10)
        private val Z = LetterNode('Z', 10)

        val letterNodeMap: Map<Char, LetterNode> = mapOf(
            ' ' to SPACE,
            'E' to E, 'A' to A, 'I' to I, 'O' to O, 'N' to N,
            'R' to R, 'T' to T, 'L' to L, 'S' to S, 'U' to U,
            'D' to D, 'G' to G,
            'B' to B, 'C' to C, 'M' to M, 'P' to P,
            'F' to F, 'H' to H, 'V' to V, 'W' to W, 'Y' to Y,
            'K' to K,
            'J' to J, 'X' to X,
            'Q' to Q, 'Z' to Z
        )

        fun createRoot(): WordTree {
            return WordTree(LetterNode('.', 0))
        }

    }

}

File: app/src/main/cpp/models/LetterNode.cpp
#include "LetterNode.h"

#include <array>

constexpr LetterNode LetterNode::SPACE{' ', 0};
constexpr LetterNode LetterNode::E{'E', 1};
constexpr LetterNode LetterNode::A{'A', 1};
constexpr LetterNode LetterNode::I{'I', 1};
constexpr LetterNode LetterNode::O{'O', 1};
constexpr LetterNode LetterNode::N{'N', 1};
constexpr LetterNode LetterNode::R{'R', 1};
constexpr LetterNode LetterNode::T{'T', 1};
constexpr LetterNode LetterNode::L{'L', 1};
constexpr LetterNode LetterNode::S{'S', 1};
constexpr LetterNode LetterNode::U{'U', 1};

constexpr LetterNode LetterNode::D{'D', 2};
constexpr LetterNode LetterNode::G{'G', 2};

constexpr LetterNode LetterNode::B{'B', 3};
constexpr LetterNode LetterNode::C{'C', 3};
constexpr LetterNode LetterNode::M{'M', 3};
constexpr LetterNode LetterNode::P{'P', 3};

constexpr LetterNode LetterNode::F{'F', 4};
constexpr LetterNode LetterNode::H{'H', 4};
constexpr LetterNode LetterNode::V{'V', 4};
constexpr LetterNode LetterNode::W{'W', 4};
constexpr LetterNode LetterNode::Y{'Y', 4};

constexpr LetterNode LetterNode::K{'K', 5};

constexpr LetterNode LetterNode::J{'J', 8};
constexpr LetterNode LetterNode::X{'X', 8};

constexpr LetterNode LetterNode::Q{'Q', 10};
constexpr LetterNode LetterNode::Z{'Z', 10};

constexpr std::array<const LetterNode*, 27> LetterNode::letterNodeMap = []() {
    std::array<const LetterNode*, 27> map = {};

    map[0] = &LetterNode::SPACE; // @ - 64 gives 0
    map['A' - 64] = &LetterNode::A;
    map['B' - 64] = &LetterNode::B;
    map['C' - 64] = &LetterNode::C;
    map['D' - 64] = &LetterNode::D;
    map['E' - 64] = &LetterNode::E;
    map['F' - 64] = &LetterNode::F;
    map['G' - 64] = &LetterNode::G;
    map['H' - 64] = &LetterNode::H;
    map['I' - 64] = &LetterNode::I;
    map['J' - 64] = &LetterNode::J;
    map['K' - 64] = &LetterNode::K;
    map['L' - 64] = &LetterNode::L;
    map['M' - 64] = &LetterNode::M;
    map['N' - 64] = &LetterNode::N;
    map['O' - 64] = &LetterNode::O;
    map['P' - 64] = &LetterNode::P;
    map['Q' - 64] = &LetterNode::Q;
    map['R' - 64] = &LetterNode::R;
    map['S' - 64] = &LetterNode::S;
    map['T' - 64] = &LetterNode::T;
    map['U' - 64] = &LetterNode::U;
    map['V' - 64] = &LetterNode::V;
    map['W' - 64] = &LetterNode::W;
    map['X' - 64] = &LetterNode::X;
    map['Y' - 64] = &LetterNode::Y;
    map['Z' - 64] = &LetterNode::Z;

    return map;
}();